litgpt/generate/base.py

```diff
@@ -24,13 +24,28 @@ def multinomial_num_samples_1(probs: torch.Tensor) -> torch.Tensor:
     return torch.multinomial(probs, num_samples=1)
 
 
-def sample(logits: torch.Tensor, temperature: float = 1.0, top_k: Optional[int] = None) -> torch.Tensor:
+def sample(
+    logits: torch.Tensor, temperature: float = 1.0, top_k: Optional[int] = None, top_p: Optional[float] = None
+) -> torch.Tensor:
     logits = logits[0, -1]
     # optionally crop the logits to only the top k options
     if top_k is not None:
         v, i = torch.topk(logits, min(top_k, logits.size(-1)))
         # do not use `torch.where` as in nanogpt because it will repeat top-k collisions
         logits = torch.full_like(logits, float("-inf")).scatter_(-1, i, v)
+    # optionally crop the logits to smallest set of logits with a cumulative probability above top_p
+    if top_p is not None and top_p < 1.0:
```

Comment:
It would look nicer if it was refactored into a separate function. You can for example look at the Mistral implementation: https://github.com/mistralai/mistral-src/blob/8598cf582091a596671be31990448e0620017851/main.py#L12-L27

Reply:
Sure. Do we also want to change the temperature scaling to be applied first as in HF? In the end one would have:

1. Temperature scaling
2. Top-k 
3. Top-p

Reply:
The order of (1) and (2) doesn't matter, right? It's only important that scaling is done before top-p. Having top-k first would also come with a tiny reduction in performance since you are trimming off the excess computations.

Reply:
Yeah, you're right! I've updated the code with your suggestions